src/rate_limiter/src/lib.rs

```diff
@@ -471,6 +519,52 @@ impl Default for RateLimiter {
     }
 }
 
+/// State for saving a RateLimiter.
+#[derive(Versionize)]
+pub struct RateLimiterState {
+    ops: Option<TokenBucketState>,
+    bandwidth: Option<TokenBucketState>,
+    timer_active: bool,
+}
+
+impl Persist for RateLimiter {
+    type State = RateLimiterState;
+    type ConstructorArgs = ();
+    type Error = io::Error;
+
+    fn save(&self) -> Self::State {
+        RateLimiterState {
+            ops: self.ops.as_ref().map(|ops| ops.save()),
+            bandwidth: self.bandwidth.as_ref().map(|bw| bw.save()),
+            timer_active: self.timer_active,
+        }
+    }
+
+    fn restore(_: Self::ConstructorArgs, state: &Self::State) -> Result<Self, Self::Error> {
+        let mut rate_limiter = RateLimiter {
+            // Safe to unwrap because TokenBucket::restore doesn't return errors.
```

Comment:
More specifically, errors should be returned when serializing the state using semantic serialization. From this perspective it kind of makes sense that restore would return Self instead of Result.